### PR TITLE
Simple Rust driver that touches real hardware: PR 2/6

### PR DESCRIPTION
**Pull request by TheSven73**
#### Simple Rust driver that touches real h/w: step 2 of 6

@alex and @ojeda suggested that #254 should be split into smaller PRs, so they can be reviewed and merged individually.

Roadmap:

- [x] platform_device
- [ ] **devicetree (of) matching (you are here)**
- [ ] probe
- [ ] DrvData
- [ ] regmap/iomem
- [ ] finally the actual hwrng driver

### CODE REVIEW

File: rust/kernel/of.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/// Wraps a kernel Open Firmware / devicetree match table.
+///
+/// Rust drivers may create this structure to match against devices
+/// described in the devicetree.
+///
+/// # Invariants
+///
+/// The last array element is always filled with zeros (the default).
+pub struct OfMatchTable(Box<[bindings::of_device_id; 2]>);
```
Comment (TheSven73) [inline rust/kernel/of.rs]: I tried to support variable-size match tables by using const generics here, eg:
```rust
pub struct OfMatchTable<const N: usize>(Box<[bindings::of_device_id; N+1]>);
```
and everything appeared to work great, until building rustdoc. That failed with the internal compiler error described [here](https://github.com/rust-lang/rust/issues/77647).

Comment (wedsonaf) [inline rust/kernel/of.rs]: For public APIs, we're trying to avoid dependencies on `bindings`. In fact, eventually we want to make `bindings` private.

So I suggest a newtype here that wraps `bindings::of_device_id`. Or even some Rust struct that is easy to use by clients, accompanied by a (private) `const` function that converts that into an array of `bindings::of_device_id` -- which is what we do with file operations, for example.

File: rust/kernel/platdev.rs [hunk 2]
```diff
@@ -49,13 +51,20 @@ impl Registration {
...
         this.pdrv.remove = Some(remove_callback);
         // SAFETY:
         //   - `this.pdrv` lives at least until the call to `platform_driver_unregister()` returns.
         //   - `name` pointer has static lifetime.
         //   - `module.0` lives at least as long as the module.
         //   - `probe()` and `remove()` are static functions.
+        //   - `of_match_tbl` is either:
+        //      - a raw pointer created from a heap structure which is never deallocated,
+        //        therefore it lives at least as long as the module, or
+        //      - null.
```
Comment (wedsonaf) [inline rust/kernel/platdev.rs:72]: I think we also need to support `'static` pointers, which would be created by `const` functions. Unless, of course, there is a technical reason that prevents us from doing so.

File: rust/kernel/of.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        // element, therefore that final element will be filled with zeros,
+        // and the invariant above will hold.
+        Ok(Self(tbl))
+    }
+
+    /// Transforms this [`OfMatchTable`] into a raw pointer.
+    ///
+    /// The resulting raw pointer is suitable to be assigned to a
+    /// `bindings::device_driver::of_match_table`.
+    pub(crate) fn into_table_ptr(self) -> *const bindings::of_device_id {
```
Comment (wedsonaf) [inline rust/kernel/of.rs]: Are you intentionally consuming the `Box` here?

I think I understand why you'd want to do this, but I'm not a big fan of APIs that leak by design. Ideally we should have a way to recover and free this box if we unregisters the platform device.

File: rust/kernel/of.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        // CAST: the kernel C API expects a pointer to an `bindings::of_device_id`
+        // array, with the final element set to zeros, which serves as a
+        // sentinel. This is exactly what we have created, as per the invariant
+        // above, so the cast is safe.
+        self.0.into_pointer().cast()
+    }
+
+    fn new_of_device_id(compatible: &CStr<'static>) -> Result<bindings::of_device_id> {
```
Comment (wedsonaf) [inline rust/kernel/of.rs:46]: If feels like the length check could be done at compile time so that we reject bad strings as early as possible. We may need a macro + const fn + static_assert combination but I think it's doable. (And likely preferable if we can make the ergonomics good enough.)

File: rust/kernel/of.rs (hunk 1 above)
Comment (TheSven73) [reply rust/kernel/of.rs]: Hi @wedsonaf thank you so much for the review, really appreciate it !

I'm not sure I understand your feedback - the `bindings::` member inside the public `OfMatchTable` is private, so clients outside `of.rs` don't know, see or care that it contains a (private) `bindings::` type?

My apologies if I have misunderstood. Could you provide an example of what your concern is? That usually helps me "see the light".

File: rust/kernel/platdev.rs (hunk 2 above)
Comment (TheSven73) [reply rust/kernel/platdev.rs:72]: I agree 100% that the structure should be `'static`, `const`, and created at build time - I even indicated that in the PR's first commit message.

If we can go that route, your remaining concerns disappear, correct?
- match table would be a`'static` pointer
- we would not have to use a `Box` (ugh that looked bad - I know !)
- length checks would happen at build time

In this PR's v2, I can get rid of the ugly `Box` issue, and make sure that it gets cleaned up properly, without resulting in a memory leak.

The problem is - I have no idea (yet!) how to create the table statically at build time.

Could I convince you to defer this issue until a later PR?
Can't get the perfect get in the way of the good?

Comment (TheSven73) [reply rust/kernel/platdev.rs:72]: PS I tried using const generics when creating the match table structure. That's probably one of the things we'll need to create the table at build time? I ran into internal compiler errors, which is kind of an indication we're in uncharted territory here...

File: rust/kernel/of.rs (hunk 1 above)
Comment (wedsonaf) [reply rust/kernel/of.rs]: Oh, my bad. For some reason I read this as a type alias instead of a new type. I even suggested that you do exactly what you're already doing.

File: rust/kernel/platdev.rs (hunk 2 above)
Comment (wedsonaf) [reply rust/kernel/platdev.rs:72]: Yes, if we don't have the box issue, I think we could get this is and improve on it later.

**Comment by ksquirrel** [comment]
Review of `453d7ea3f792`:
  - ✔️ Commit [`cc2c87b`](https://github.com/Rust-for-Linux/linux/commit/cc2c87b64e2eadfd9461096939617234efcedf66): Looks fine!
  - ✔️ Commit [`a8370a4`](https://github.com/Rust-for-Linux/linux/commit/a8370a42a1adc3648ce2a3be7392d0fc13417367): Looks fine!
  - ✔️ Commit [`453d7ea`](https://github.com/Rust-for-Linux/linux/commit/453d7ea3f792c0718a13deec5edc46d546227977): Looks fine!

**Comment by ksquirrel** [comment]
Review of `a0d9af807ef5`:
  - ✔️ Commit [`a2bbb25`](https://github.com/Rust-for-Linux/linux/commit/a2bbb25ccd6c6c7e55e7624ac0ff7f59a113db74): Looks fine!
  - ✔️ Commit [`f92bce2`](https://github.com/Rust-for-Linux/linux/commit/f92bce28b347db29528334f904860903b725cfc6): Looks fine!
  - ✔️ Commit [`a0d9af8`](https://github.com/Rust-for-Linux/linux/commit/a0d9af807ef5201fbf9d4121b497b2d68e524d99): Looks fine!

**Comment by TheSven73** [comment]
### v1 -> v2
wedsonaf:
- clean up of_table after `platform_driver_unregister()`

I opted to clean up the table by keeping a copy of the `PointerWrapper` raw pointer, and clean it up on `Drop`. I think that looks quite ugly though:
```rust
impl Drop for Registration {
    fn drop(&mut self) {
        if self.registered {
            // SAFETY: if `registered` is true, then `self.pdev` was registered
            // previously, which means `platform_driver_unregister` is always
            // safe to call.
            unsafe { bindings::platform_driver_unregister(&mut self.pdrv) }
            if let Some(ptr) = self.of_table {
                // SAFETY: `ptr` came from an `OfMatchTable`.
                let tbl = unsafe { OfMatchTable::from_pointer(ptr) };
                drop(tbl);
            }
        }
    }
}
```

My immediate instinct would be to find some general solution. Maybe we can convert the `PointerWrapper` into a `PointerHolder` which allows the raw pointer to be retrieved, while keeping ownership of the `PointerWrapper` ? Something along the following lines:
```rust
/// Used to retain ownership of a [`PointerWrapper`] after the `into_pointer()` call.
pub struct PointerHolder<T: PointerWrapper> {
    ptr: *const c_types::c_void,
    _phantom: PhantomData<T>,
}

impl<T: PointerWrapper> Drop for PointerHolder<T> {
    fn drop(&mut self) {
        let obj = unsafe { T::from_pointer(self.ptr) };
        drop(obj);
    }
}

impl<T: PointerWrapper> PointerHolder<T> {
    pub fn as_ref(&self) -> *const c_types::c_void {
        self.ptr
    }
}

impl<T: PointerWrapper> From<T> for PointerHolder<T> {
    fn from(wrapper: T) -> Self {
        PointerHolder {
            ptr: wrapper.into_pointer(),
            _phantom: PhantomData,
        }
    }
}

```
However, we are planning to make this table `static const`, at which point all of this goes away. So probably not worth generalizing at this point. Maybe we can revisit this when we encounter our first dynamically allocated kernel object that contains another dynamically allocated kernel object?

Apologies for the extended brain dump :)

**Comment by ksquirrel** [comment]
Review of `c02591f2b011`:
  - ✔️ Commit [`fb3fdcb`](https://github.com/Rust-for-Linux/linux/commit/fb3fdcbabdf679cca0006b23fcbcf03c9a610f40): Looks fine!
  - ✔️ Commit [`6e4043a`](https://github.com/Rust-for-Linux/linux/commit/6e4043a7a147491838d2ae8b7e2b045cd315a51e): Looks fine!
  - ✔️ Commit [`c02591f`](https://github.com/Rust-for-Linux/linux/commit/c02591f2b011e6544db51b47906a4f3e339735ac): Looks fine!

**Comment by TheSven73** [comment]
### v1 -> v2
wedsonaf:
- clean up of_table after `platform_driver_unregister()`

TheSven73:
- implement `PointerWrapper` for `OfMatchTable` which is idiomatic for obtaining kernel pointers to store in kernel structures

I opted to clean up the table by keeping a copy of the `PointerWrapper` raw pointer, and clean it up on `Drop`. I think that looks quite ugly though:
```rust
impl Drop for Registration {
    fn drop(&mut self) {
        if self.registered {
            // SAFETY: if `registered` is true, then `self.pdev` was registered
            // previously, which means `platform_driver_unregister` is always
            // safe to call.
            unsafe { bindings::platform_driver_unregister(&mut self.pdrv) }
        }
        if let Some(ptr) = self.of_table {
            // SAFETY: `ptr` came from an `OfMatchTable`.
            let tbl = unsafe { OfMatchTable::from_pointer(ptr) };
            drop(tbl);
        }
    }
}
```

My immediate instinct would be to find some general solution. Maybe we can convert the `PointerWrapper` into a `PointerHolder` which allows the raw pointer to be retrieved, while keeping ownership of the `PointerWrapper` ? Something along the following lines:
```rust
/// Used to retain ownership of a [`PointerWrapper`] after the `into_pointer()` call.
pub struct PointerHolder<T: PointerWrapper> {
    ptr: *const c_types::c_void,
    _phantom: PhantomData<T>,
}

impl<T: PointerWrapper> Drop for PointerHolder<T> {
    fn drop(&mut self) {
        let obj = unsafe { T::from_pointer(self.ptr) };
        drop(obj);
    }
}

impl<T: PointerWrapper> PointerHolder<T> {
    pub fn as_ref(&self) -> *const c_types::c_void {
        self.ptr
    }
}

impl<T: PointerWrapper> From<T> for PointerHolder<T> {
    fn from(wrapper: T) -> Self {
        PointerHolder {
            ptr: wrapper.into_pointer(),
            _phantom: PhantomData,
        }
    }
}

/// A registration of a platform device.
#[derive(Default)]
pub struct Registration {
    registered: bool,
-   of_table: Option<*const c_types::c_void>,
+   // OfMatchTable automatically cleaned up on Drop
+   of_table: Option<PointerHolder<OfMatchTable>>,
    pdrv: bindings::platform_driver,
    _pin: PhantomPinned,
}
```
However, we are planning to make this table `static const`, at which point all of this goes away. So probably not worth generalizing at this point. Maybe we can revisit this when we encounter our first dynamically allocated kernel object that contains another dynamically allocated kernel object?

Apologies for the extended brain dump :)

Edit: clarified `PointerHolder` riff.
Edit2: mention that v2 implements `PointerWrapper` for `OfMatchTable`. Sorry for the noise.

**Comment by nbdd0121** [comment]
@TheSven73 have you checked the `CBoundedStr<N>` API proposed in #258? It has a constant upper bound for length `N`, can be created safely in compile time, and provide a method to expand into a fixed length char array.

File: rust/kernel/platdev.rs [hunk 5]
```diff
@@ -18,6 +21,7 @@ use core::{marker::PhantomPinned, pin::Pin};
 #[derive(Default)]
 pub struct Registration {
     registered: bool,
+    of_table: Option<*const c_types::c_void>,
```
Comment (soruh) [inline rust/kernel/platdev.rs:24]: I think this can be an `Option<NonNull<c_types::c_void>>` because as far as I can see all pointers it could contain ultimately come from a `Box` allocation. (This would safe at least one byte for the `Option` discriminant)

Comment (TheSven73) [reply rust/kernel/platdev.rs:24]: As far as I can see, creating a `NonNull` from a `ptr` requires one additional `unwrap` or `Err()` return, which would introduce one more `// PANIC:` annotation?
```rust
pub fn new(ptr: *mut T) -> Option<NonNull<T>>
```
Also its docs say `but is ultimately more dangerous to use because of its additional properties` ?
So not sure if the pros outweigh the cons here.
But thank you for introducing me to `NonNull` ! It's basically a covariant pointer type, I didn't even know that this existed. That's great to know.

Comment (soruh) [reply rust/kernel/platdev.rs:24]: > As far as I can see, creating a NonNull from a ptr requires one additional unwrap or Err() return, which would introduce one more // PANIC: annotation?

correct but since we create that pointer by using a `Box::into_raw` we can use `NonNull::from(Box::leak(...))` instead.

This would of course require using a `NonNull` everywhere where that pointer was used but I think that is manageable/mostly handled by the compiler in this case.

> Also its docs say but is ultimately more dangerous to use because of its additional properties?
So not sure if the pros outweigh the cons here.

I think that message is mostly for people new to Rust being told to use `NonNull` and attempting to use it like a normal C/raw pointer.

From my experience working with `NonNull`s it is generally easier to reason that code is safe when dereferencing (since you don't have to argue that a pointer can't be `NULL` everywhere)
but harder when creating a pointer. e.g. pointer arithmetic becomes very verbose since you now need to argue that your arithmetic can't make the pointer `NULL` (which I think is a good thing anyways).

TL;DR using `NonNull`s made from references is easier, doing pointer magic with them is harder / more error prone.

(I do of course understand if you still think this isn't worth the risk)

> But thank you for introducing me to NonNull ! It's basically a covariant pointer type, I didn't even know that this existed. That's great to know.

You're welcome :)

Comment (TheSven73) [reply rust/kernel/platdev.rs:24]: I suspect that in this case (a one-off use inside `Registration`), the light is probably not worth the candle. But, `NonNull<T>`  might be extremely useful as part of the implementation detail of more general, core abstractions. The more such an abstraction is re-used, the more we can accept verbosity and "tricky" guarantees inside of it.

I really appreciate your knowledge and expertise !

Comment (soruh) [reply rust/kernel/platdev.rs:24]: I had assumed that the `Registration` would live for the whole lifetime of the module so saving up to 8 bytes for basically free would be worth it but I agree that in this case it probably doesn't matter that much.

> I really appreciate your knowledge and expertise!

I'm glad I can at least contribute something on the Rust usage side since I have very little prior experience with linux kernel code.

Comment (TheSven73) [reply rust/kernel/platdev.rs:24]: Ah, I see what you mean, good point, I hadn't fully understood the space savings yet. I don't believe the 8 byte savings per `Registration` is significant here however: there is only a single `Registration` instance per driver. All devices (driver instances) share a single `Registration` instance. So if we had 1000 hardware random-number generators in our system, we'd save only 8 bytes, not 8000. At least, that's my understanding of the situation.

Comment (soruh) [reply rust/kernel/platdev.rs:24]: > But, `NonNull<T>` might be extremely useful as part of the implementation detail of more general, core abstractions.

Would it maybe make sense to encourage using NonNull by providing a helper method like this:
```rust
/// Attempts to allocate `x` on the heap using a `Box` and provides a pointer to that allocation.
fn bikeshed_alloc<T>(x: T) -> Result<NonNull<T>> {
    NonNull::from(Box::leak(Box::try_new(x)?))
}
```

This would remove most of the complexity around using `NonNull`, deduplicate a lot of the `Box::into_raw(Box::try_new(...)?)` logic and would of course still allow people to use `.as_ptr()` to get a "normal" raw pointer if they need to e.g. do arithmetic with it.

EDIT: I just realized it might make more sense for this comment to be it's own issue as to not pollute discussion around this PR too much, I'll open one in a bit.

Comment (TheSven73) [reply rust/kernel/platdev.rs:24]: Not sure. `Box`es aren't the only structures that can convert `into()` shared kernel pointers! `PointerWrapper` is the trait that's supposed to be implemented by objects that can go from/into shared pointers. Currently implemented for `Box`, `Arc`, `Pin` and `Ref`.

Comment (soruh) [reply rust/kernel/platdev.rs:24]: Ah I had not yet encountered `PointerWrapper`. I agree that it doesn't really make sense to provide a helper parallel to that.

Comment (TheSven73) [reply rust/kernel/platdev.rs:24]: `NonNull<T>` may make sense in some of the code surrounding `PointerWrapper - when there's a case to refine that abstraction.

Comment (nbdd0121) [reply rust/kernel/platdev.rs:24]: `NonNull` is generally preferred in data structures wherever a pointer is sure to be not null. `Option<NonNull<T>>` can be used if a pointer is nullable. Note that `Option<NonNull<T>>` is even FFI-safe -- it is treated exactly as `*mut T` for ABI purporses.

Doing so forces the user to consider about the nullability of a pointer, so it can prevent null dereference even in unsafe code. The only exception are probably local variables, as raw pointers have more helper function to do arithmetic on.

Comment (soruh) [reply rust/kernel/platdev.rs:24]: Hmm, I think I'll take this as an opportunity to familiarize myself with the code base and look for opportunities where `NonNull` would be applicable and an actual improvement.

File: rust/kernel/of.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+/// # Invariants
+///
+/// The final array element is always filled with zeros (the default).
+pub struct OfMatchTable(InnerTable);
+
+impl OfMatchTable {
+    /// Creates a [`OfMatchTable`] from a single `compatible` string.
+    pub fn new(compatible: &CStr<'static>) -> Result<Self> {
+        let tbl: InnerTable = Box::try_new([
+            Self::new_of_device_id(compatible)?,
```
Comment (Skallwar) [inline rust/kernel/of.rs:37]: Shouldn't this allow for multiple match strings? Isn't the point of the match table to allow for multiple `of_device_id`? We've used it in a driver, and apart from the array needing to be NULL-terminated, it can have any number of elements, not necessarily just one with a comma-separated string. But we might be wrong.

File: rust/kernel/of.rs (hunk 4 above)
Comment (Skallwar) [reply rust/kernel/of.rs:46]: ``.len()`` is ``const`` so replacing line 48 with a static_assert should work out of the box

File: rust/kernel/platdev.rs [hunk 7]
```diff
@@ -85,5 +100,10 @@ impl Drop for Registration {
             // safe to call.
             unsafe { bindings::platform_driver_unregister(&mut self.pdrv) }
         }
+        if let Some(ptr) = self.of_table {
+            // SAFETY: `ptr` came from an `OfMatchTable`.
+            let tbl = unsafe { OfMatchTable::from_pointer(ptr) };
+            drop(tbl);
```
Comment (Skallwar) [inline rust/kernel/platdev.rs:106]: I don't understand why we need to drop it

File: rust/kernel/platdev.rs [hunk 8]
```diff
@@ -49,13 +54,22 @@ impl Registration {
             return Err(Error::EINVAL);
         }
         this.pdrv.driver.name = name.as_ptr() as *const c_types::c_char;
+        if let Some(tbl) = of_match_table {
+            let ptr = tbl.into_pointer();
+            this.of_table = Some(ptr);
+            this.pdrv.driver.of_match_table = ptr.cast();
+        }
```
Comment (Skallwar) [inline rust/kernel/platdev.rs:61]: Maybe ``this.of_table`` should be a ``Option<OfMatchTable>``. I don't understand the need of a raw pointer here. This could fix the ``drop`` later on ?

File: rust/kernel/of.rs (hunk 6 above)
Comment (TheSven73) [reply rust/kernel/of.rs:37]: Yes, absolutely! That's planned for a follow-up PR. In fact, we want the table to be generated `static const` but we don't  know how yet. I tried generating a variable-size table (for multiple match strings) using const generics, but I ran into an internal rustc error...

File: rust/kernel/platdev.rs (hunk 7 above)
Comment (TheSven73) [reply rust/kernel/platdev.rs:106]: The `drop(tbl)` is completely redundant, `tbl` would get dropped even without it. It's  just there to indicate to the reader that we are dropping the table there.

File: rust/kernel/platdev.rs (hunk 8 above)
Comment (TheSven73) [reply rust/kernel/platdev.rs:61]: It's because structures that are shared with kernel structs using raw pointers, use the `PointerWrapper` abstraction/trait. This trait ensures that the struct is consumed when turned into a kernel pointer. After the call to `into_pointer()`, there is no `tbl` object left to store anywhere !

**Comment by nbdd0121** [comment]
I suppose that you could use a similar approach to the one used in #273 for CStr.
```rust
#[repr(transparent)]
struct OfDeviceId(bindings::of_device_id);

impl OfDeviceId {
    pub const fn nul() -> Self {
        // give a nul
    }

    pub const fn new(/* blah */) -> Self {
        // construct
    }
}

#[repr(transparent)]
struct OfMatchTable([OfDeviceId]);

impl OfMatchTable {
    pub const fn from_slice_with_nul(arr: &[OfDeviceId]) -> &Self {
        // Verify that none of non-last entries are NUL
        // Verify that last entry of arr is indeed NUL
        unsafe { Self::from_slice_with_nul_unchecked(arr) }
    }
    
    pub const unsafe fn from_slice_with_nul_unchecked(arr: &[OfDeviceId]) -> &Self {
        &*(arr as *const [OfDeviceId] as *const Self)
    }
}
```
then people could use
```rust
const TBL: &'static OfMatchTable = OfMatchTable::from_slice_with_nul(&[OfDeviceId::new(/*blah*/), OfDeviceId::nul()]);
```

**Comment by TheSven73** [comment]
@nbdd0121 that's a cool idea actually !

**Comment by TheSven73** [comment]
I'd prefer to get this merged without `const fn`, `static` or variable-size match tables for now, if possible. Can't get the perfect get in the way of the good. We can improve in a follow-up PR?

File: rust/kernel/of.rs (hunk 6 above)
Comment (Skallwar) [reply rust/kernel/of.rs:37]: Ok nice

File: rust/kernel/platdev.rs (hunk 7 above)
Comment (Skallwar) [reply rust/kernel/platdev.rs:106]: I see. Imao, it's not pretty :sweat_smile:

Comment (TheSven73) [reply rust/kernel/platdev.rs:106]: Hehe I agree... I think I used the word [ugly](https://github.com/Rust-for-Linux/linux/pull/276#issuecomment-844475911) myself :)

**Comment by TheSven73** [comment]
@wedsonaf in case you are unsatisfied with v2, feel free to ask for a "reboot". I can handle constructive criticism, and have plenty of alternative approaches up my sleeve :)

**Comment by wedsonaf** [comment]
> @wedsonaf in case you are unsatisfied with v2, feel free to ask for a "reboot". I can handle constructive criticism, and have plenty of alternative approaches up my sleeve :)

Looking at it now. Give me a few minutes. :)

**Comment by TheSven73** [comment]
No time pressure here at all. Please do take as much time as you need!

File: rust/kernel/of.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+///
+/// # Invariants
+///
+/// The final array element is always filled with zeros (the default).
+pub struct OfMatchTable(InnerTable);
+
+impl OfMatchTable {
+    /// Creates a [`OfMatchTable`] from a single `compatible` string.
+    pub fn new(compatible: &CStr<'static>) -> Result<Self> {
+        let tbl: InnerTable = Box::try_new([
```
Comment (wedsonaf) [inline rust/kernel/of.rs:36]: nit: do we need the type (`InnerTable`) explicitly given here?

**Review by wedsonaf — APPROVED** [review]
LGTM.

Let's merge this so that we can make progress on the driver.

We can always iterate on the wrinkles that don't affect safety.

File: rust/kernel/of.rs (hunk 9 above)
Comment (TheSven73) [reply rust/kernel/of.rs:36]: You are right - there's no need. I don't like unnecessary explicit type annotations either. I'll do a follow-up PR to eliminate it.
